_posts/2016-06-30-state-in-scalable-architectures.md

```diff
@@ -0,0 +1,173 @@
+
+If you went through a similar train of thought like myself these concepts appeared for sure: mutability, concurrency, isolation and scope. Most of them has to do with the level of abstraction or the strategy to handle state, not with the definition itself. Let's try to extract the essential from [Wikipedia's definition](https://www.wikiwand.com/en/State_(computer_science)):
+
+>>The state of a digital logic circuit or computer program is a technical term for all the stored information, at a given instant in time, to which the circuit or program has access.
+
+Data at a given instant in time. Neither behaviour, nor abstractions around it, but pure data.
+
+[Out of the tar pit paper](https://github.com/papers-we-love/papers-we-love/blob/master/design/out-of-the-tar-pit.pdf) has powerful insights about state from a more radical concept than scalability: complexity.
+
+>> One final point is that the type of complexity we are discussing in this paper is that which makes large systems hard to understand. It is this that causes us to expend huge resources in creating and maintaining such systems. This type of complexity has nothing to do with complexity theory — the branch of computer science which studies the resources consumed by a machine executing a program.
+
+## State at service level
+
+Through this post we'll see how simplicity and scalability aligns blissfully when handling state. Let's start with approaches to handle state at service level. We'll move then into distributed systems integration.
+
+### Global State
+
+This is the loosest approach, e.g. `public static` non final variables in Java. How can we informal reason about code that uses global variables? We need to inspect all the code that is mutating that variable, and those lines of code might live in completely unrelated parts of the system. Having to reason about plenty of pieces at the same time is inefficient and dangerous. Global state forces us to do black box testing as the pieces of our system are coupled through global state.
+
+A priori simple lines of code get quickly 'contaminated' by the complexity derived of state:
+
+>> The key problem is that a test (of any kind) on a system or component that is in one particular state tells you nothing at all about the behaviour of that system or component when it happens to be in another state.
+
+### Object Oriented Programming State
+
+>> In most forms of object-oriented programming (OOP) an object is seen as consisting of some state together with a set of procedures for accessing and manipulating that state.
+
+As [Uncle Bob](https://www.wikiwand.com/en/Robert_Cecil_Martin) said once, programming paradigms succeeds, paradoxically, by removing power. Java succeed because it removed from developers the power to manage memory. OOP strongly suggests that code that mutates state should be colocated with that state. That guarantee, along preconditions and postconditions, makes code much easier to reason about and test.
```

Comment:
programming paradigms succeed, paradoxically, ...

Java succeed**s** because it removed...